#### 05 - Explainability Score

**Measuring token-level importance using SHAP attribution over 25 SST-2 sentences per model. A focused model assigns high importance to fewer, meaningful tokens rather than spreading attribution uniformly. Explainability score is derived from the Gini coefficient of SHAP values, higher concentration = more explainable.**

In [2]:
import sys
import os
import yaml
import pandas as pd
from dataclasses import asdict
import json
# sys.path.append(os.path.abspath(".."))
# from src import explainability_scoring
import explainability_scoring
import importlib
importlib.reload(explainability_scoring)
from datasets import load_dataset
import warnings
warnings.filterwarnings('ignore')

In [3]:
with open("config.yaml", "r") as f:
    config = yaml.safe_load(f)
models=config['models']['list']

In [4]:
# reusing SST-2 validation set
sst2= load_dataset("sst2", split="validation")
sentences = [row["sentence"] for row in sst2.select(range(25))]
print(f"loaded {len(sentences)} sentences")
print(f"example: {sentences[0]}")

loaded 25 sentences
example: it 's a charming and often affecting journey . 


In [5]:
results = [explainability_scoring.evaluate_explainability(m, sentences) for m in models]
print(f"\ndone. evaluated {len(results)} models.")


Evaluating: TinyLlama/TinyLlama-1.1B-Chat-v1.0


config.json:   0%|          | 0.00/608 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/551 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.20G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/201 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

computing SHAP:   0%|          | 0/25 [00:00<?, ?it/s]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:03, 63.76s/it]               
computing SHAP:   4%|▍         | 1/25 [01:03<25:30, 63.76s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:46, 46.59s/it]               
computing SHAP:   8%|▊         | 2/25 [01:50<20:34, 53.66s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:17, 77.62s/it]               
computing SHAP:  12%|█▏        | 3/25 [03:07<23:41, 64.60s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:21, 81.93s/it]               
computing SHAP:  16%|█▌        | 4/25 [04:29<25:00, 71.45s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:51, 51.49s/it]               
computing SHAP:  20%|██        | 5/25 [05:21<21:25, 64.25s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:18, 78.68s/it]               
computing SHAP:  24%|██▍       | 6/25 [06:40<21:54, 69.16s/it]

  0%|          | 0/56 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:36, 36.37s/it]               
computing SHAP:  28%|██▊       | 7/25 [07:16<17:31, 58.44s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:02, 62.86s/it]               
computing SHAP:  32%|███▏      | 8/25 [08:19<16:57, 59.85s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:05, 65.96s/it]               
computing SHAP:  36%|███▌      | 9/25 [09:25<16:28, 61.76s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:16, 76.99s/it]               
computing SHAP:  40%|████      | 10/25 [10:42<16:36, 66.46s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:11, 71.19s/it]               
computing SHAP:  44%|████▍     | 11/25 [11:53<15:50, 67.91s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:21, 81.09s/it]               
computing SHAP:  48%|████▊     | 12/25 [13:14<15:35, 71.93s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:10, 70.93s/it]               
computing SHAP:  52%|█████▏    | 13/25 [14:25<14:19, 71.63s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:15, 75.81s/it]               
computing SHAP:  56%|█████▌    | 14/25 [15:41<13:21, 72.89s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:18, 78.33s/it]               
computing SHAP:  60%|██████    | 15/25 [16:59<12:25, 74.54s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:27, 87.31s/it]               
computing SHAP:  64%|██████▍   | 16/25 [18:27<11:45, 78.38s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:14, 74.73s/it]               
computing SHAP:  68%|██████▊   | 17/25 [19:41<10:18, 77.28s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:35, 95.19s/it]               
computing SHAP:  72%|███████▏  | 18/25 [21:16<09:38, 82.67s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:49, 49.79s/it]               
computing SHAP:  76%|███████▌  | 19/25 [22:06<07:16, 72.79s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:26, 86.07s/it]               
computing SHAP:  80%|████████  | 20/25 [23:32<06:23, 76.78s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:27, 87.51s/it]               
computing SHAP:  84%|████████▍ | 21/25 [25:00<05:20, 80.00s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:01, 61.03s/it]               
computing SHAP:  88%|████████▊ | 22/25 [26:01<03:42, 74.31s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:49, 49.46s/it]               
computing SHAP:  92%|█████████▏| 23/25 [26:50<02:13, 66.86s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:13, 73.40s/it]               
computing SHAP:  96%|█████████▌| 24/25 [28:04<01:08, 68.82s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:11, 71.17s/it]               
computing SHAP: 100%|██████████| 25/25 [29:15<00:00, 70.22s/it]


  explainability: 0.33  |  mean gini: 0.33  |  avg top tokens: 1.48

Evaluating: microsoft/phi-1_5


Loading weights:   0%|          | 0/341 [00:00<?, ?it/s]

computing SHAP:   0%|          | 0/25 [00:00<?, ?it/s]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:58, 59.00s/it]               
computing SHAP:   4%|▍         | 1/25 [00:59<23:36, 59.00s/it]

  0%|          | 0/56 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:28, 28.25s/it]               
computing SHAP:   8%|▊         | 2/25 [01:27<15:41, 40.92s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:18, 78.76s/it]               
computing SHAP:  12%|█▏        | 3/25 [02:46<21:20, 58.20s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:12, 72.86s/it]               
computing SHAP:  16%|█▌        | 4/25 [03:58<22:23, 63.99s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:01, 61.69s/it]               
computing SHAP:  20%|██        | 5/25 [05:00<21:03, 63.16s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:17, 77.85s/it]               
computing SHAP:  24%|██▍       | 6/25 [06:18<21:34, 68.15s/it]

  0%|          | 0/30 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:19, 19.80s/it]               
computing SHAP:  28%|██▊       | 7/25 [06:38<15:42, 52.35s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:09, 69.70s/it]               
computing SHAP:  32%|███▏      | 8/25 [07:47<16:23, 57.87s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:15, 75.57s/it]               
computing SHAP:  36%|███▌      | 9/25 [09:03<16:54, 63.41s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:53, 113.40s/it]              
computing SHAP:  40%|████      | 10/25 [10:56<19:42, 78.84s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:29, 89.25s/it]               
computing SHAP:  44%|████▍     | 11/25 [12:26<19:08, 82.03s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:44, 104.90s/it]              
computing SHAP:  48%|████▊     | 12/25 [14:11<19:16, 88.99s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:23, 83.61s/it]               
computing SHAP:  52%|█████▏    | 13/25 [15:34<17:28, 87.36s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:27, 87.09s/it]               
computing SHAP:  56%|█████▌    | 14/25 [17:01<16:00, 87.28s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:11, 71.19s/it]               
computing SHAP:  60%|██████    | 15/25 [18:12<13:44, 82.43s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:35, 95.39s/it]               
computing SHAP:  64%|██████▍   | 16/25 [19:48<12:57, 86.33s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:21, 81.73s/it]               
computing SHAP:  68%|██████▊   | 17/25 [21:10<11:19, 84.95s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:34, 94.28s/it]               
computing SHAP:  72%|███████▏  | 18/25 [22:44<10:14, 87.76s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:57, 57.31s/it]               
computing SHAP:  76%|███████▌  | 19/25 [23:41<07:51, 78.61s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:21, 81.98s/it]               
computing SHAP:  80%|████████  | 20/25 [25:03<06:38, 79.63s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:35, 95.14s/it]               
computing SHAP:  84%|████████▍ | 21/25 [26:38<05:37, 84.28s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:02, 62.64s/it]               
computing SHAP:  88%|████████▊ | 22/25 [27:41<03:53, 77.79s/it]

  0%|          | 0/72 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:51, 51.95s/it]               
computing SHAP:  92%|█████████▏| 23/25 [28:33<02:20, 70.04s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:05, 65.24s/it]               
computing SHAP:  96%|█████████▌| 24/25 [29:38<01:08, 68.60s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:10, 70.71s/it]               
computing SHAP: 100%|██████████| 25/25 [30:49<00:00, 73.98s/it]


  explainability: 0.6119  |  mean gini: 0.6119  |  avg top tokens: 1.6

Evaluating: Qwen/Qwen2-0.5B


config.json:   0%|          | 0.00/661 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/988M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/290 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/138 [00:00<?, ?B/s]

computing SHAP:   0%|          | 0/25 [00:00<?, ?it/s]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:24, 24.25s/it]               
computing SHAP:   4%|▍         | 1/25 [00:24<09:41, 24.25s/it]

  0%|          | 0/56 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:11, 11.35s/it]               
computing SHAP:   8%|▊         | 2/25 [00:35<06:23, 16.66s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:22, 22.71s/it]               
computing SHAP:  12%|█▏        | 3/25 [00:58<07:07, 19.43s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:25, 25.37s/it]               
computing SHAP:  16%|█▌        | 4/25 [01:23<07:37, 21.78s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:21, 21.78s/it]               
computing SHAP:  20%|██        | 5/25 [01:45<07:15, 21.78s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:29, 29.53s/it]               
computing SHAP:  24%|██▍       | 6/25 [02:15<07:43, 24.42s/it]

  0%|          | 0/30 [00:00<?, ?it/s]

computing SHAP:  28%|██▊       | 7/25 [02:22<05:38, 18.83s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:23, 23.72s/it]               
computing SHAP:  32%|███▏      | 8/25 [02:46<05:46, 20.39s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:26, 26.94s/it]               
computing SHAP:  36%|███▌      | 9/25 [03:13<05:58, 22.44s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:33, 33.07s/it]               
computing SHAP:  40%|████      | 10/25 [03:46<06:25, 25.72s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:25, 25.42s/it]               
computing SHAP:  44%|████▍     | 11/25 [04:11<05:58, 25.63s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:30, 30.68s/it]               
computing SHAP:  48%|████▊     | 12/25 [04:42<05:53, 27.17s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:27, 27.51s/it]               
computing SHAP:  52%|█████▏    | 13/25 [05:09<05:27, 27.27s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:30, 30.80s/it]               
computing SHAP:  56%|█████▌    | 14/25 [05:40<05:11, 28.34s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:23, 23.68s/it]               
computing SHAP:  60%|██████    | 15/25 [06:04<04:29, 26.94s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:29, 29.47s/it]               
computing SHAP:  64%|██████▍   | 16/25 [06:33<04:09, 27.70s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:26, 26.02s/it]               
computing SHAP:  68%|██████▊   | 17/25 [06:59<03:37, 27.20s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:34, 34.99s/it]               
computing SHAP:  72%|███████▏  | 18/25 [07:34<03:26, 29.54s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:20, 20.97s/it]               
computing SHAP:  76%|███████▌  | 19/25 [07:55<02:41, 26.97s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:26, 26.78s/it]               
computing SHAP:  80%|████████  | 20/25 [08:22<02:14, 26.91s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:33, 33.95s/it]               
computing SHAP:  84%|████████▍ | 21/25 [08:56<01:56, 29.03s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:21, 21.19s/it]               
computing SHAP:  88%|████████▊ | 22/25 [09:17<01:20, 26.68s/it]

  0%|          | 0/72 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:19, 19.61s/it]               
computing SHAP:  92%|█████████▏| 23/25 [09:37<00:49, 24.56s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:30, 30.78s/it]               
computing SHAP:  96%|█████████▌| 24/25 [10:08<00:26, 26.42s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:23, 23.17s/it]               
computing SHAP: 100%|██████████| 25/25 [10:31<00:00, 25.25s/it]


  explainability: 0.6443  |  mean gini: 0.6443  |  avg top tokens: 1.28

Evaluating: HuggingFaceTB/SmolLM-360M


config.json:   0%|          | 0.00/725 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/831 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.45G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/290 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/111 [00:00<?, ?B/s]

computing SHAP:   0%|          | 0/25 [00:00<?, ?it/s]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:19, 19.04s/it]               
computing SHAP:   4%|▍         | 1/25 [00:19<07:37, 19.05s/it]

  0%|          | 0/72 [00:00<?, ?it/s]

computing SHAP:   8%|▊         | 2/25 [00:27<04:55, 12.85s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:18, 18.46s/it]               
computing SHAP:  12%|█▏        | 3/25 [00:46<05:39, 15.41s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:22, 22.75s/it]               
computing SHAP:  16%|█▌        | 4/25 [01:08<06:24, 18.31s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:18, 18.53s/it]               
computing SHAP:  20%|██        | 5/25 [01:27<06:07, 18.39s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:22, 22.74s/it]               
computing SHAP:  24%|██▍       | 6/25 [01:50<06:17, 19.87s/it]

  0%|          | 0/30 [00:00<?, ?it/s]

computing SHAP:  28%|██▊       | 7/25 [01:55<04:33, 15.22s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:18, 18.89s/it]               
computing SHAP:  32%|███▏      | 8/25 [02:14<04:38, 16.39s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:20, 20.10s/it]               
computing SHAP:  36%|███▌      | 9/25 [02:34<04:40, 17.55s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:22, 22.55s/it]               
computing SHAP:  40%|████      | 10/25 [02:57<04:46, 19.10s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:19, 19.09s/it]               
computing SHAP:  44%|████▍     | 11/25 [03:16<04:27, 19.10s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:24, 24.87s/it]               
computing SHAP:  48%|████▊     | 12/25 [03:41<04:31, 20.85s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:22, 22.08s/it]               
computing SHAP:  52%|█████▏    | 13/25 [04:03<04:14, 21.23s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:24, 24.34s/it]               
computing SHAP:  56%|█████▌    | 14/25 [04:27<04:03, 22.17s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:17, 17.51s/it]               
computing SHAP:  60%|██████    | 15/25 [04:45<03:27, 20.77s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:24, 24.13s/it]               
computing SHAP:  64%|██████▍   | 16/25 [05:09<03:16, 21.78s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:23, 23.41s/it]               
computing SHAP:  68%|██████▊   | 17/25 [05:32<02:58, 22.27s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:23, 23.03s/it]               
computing SHAP:  72%|███████▏  | 18/25 [05:55<02:37, 22.50s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:16, 16.59s/it]               
computing SHAP:  76%|███████▌  | 19/25 [06:12<02:04, 20.73s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:22, 22.52s/it]               
computing SHAP:  80%|████████  | 20/25 [06:34<01:46, 21.27s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:25, 25.38s/it]               
computing SHAP:  84%|████████▍ | 21/25 [07:00<01:30, 22.50s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:15, 15.22s/it]               
computing SHAP:  88%|████████▊ | 22/25 [07:15<01:00, 20.32s/it]

  0%|          | 0/72 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:14, 14.70s/it]               
computing SHAP:  92%|█████████▏| 23/25 [07:30<00:37, 18.63s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:18, 18.95s/it]               
computing SHAP:  96%|█████████▌| 24/25 [07:49<00:18, 18.73s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:18, 18.32s/it]               
computing SHAP: 100%|██████████| 25/25 [08:07<00:00, 19.50s/it]


  explainability: 0.6392  |  mean gini: 0.6392  |  avg top tokens: 1.52

Evaluating: stabilityai/stablelm-2-1_6b


config.json:   0%|          | 0.00/608 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/895 [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/784 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/3.29G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/340 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/121 [00:00<?, ?B/s]

computing SHAP:   0%|          | 0/25 [00:00<?, ?it/s]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:04, 64.11s/it]               
computing SHAP:   4%|▍         | 1/25 [01:04<25:38, 64.12s/it]

  0%|          | 0/56 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:32, 32.25s/it]               
computing SHAP:   8%|▊         | 2/25 [01:36<17:23, 45.38s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:30, 90.41s/it]               
computing SHAP:  12%|█▏        | 3/25 [03:06<24:10, 65.94s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:39, 99.20s/it]               
computing SHAP:  16%|█▌        | 4/25 [04:46<27:40, 79.08s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:12, 72.96s/it]               
computing SHAP:  20%|██        | 5/25 [05:58<25:37, 76.87s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:47, 107.91s/it]              
computing SHAP:  24%|██▍       | 6/25 [07:46<27:41, 87.43s/it]

  0%|          | 0/30 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:26, 26.59s/it]               
computing SHAP:  28%|██▊       | 7/25 [08:13<20:15, 67.54s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:34, 94.66s/it]               
computing SHAP:  32%|███▏      | 8/25 [09:48<21:34, 76.17s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:36, 96.64s/it]               
computing SHAP:  36%|███▌      | 9/25 [11:24<22:01, 82.57s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:47, 107.03s/it]              
computing SHAP:  40%|████      | 10/25 [13:11<22:31, 90.13s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:25, 85.76s/it]               
computing SHAP:  44%|████▍     | 11/25 [14:37<20:43, 88.79s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:38, 98.99s/it]               
computing SHAP:  48%|████▊     | 12/25 [16:16<19:54, 91.90s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:28, 88.07s/it]               
computing SHAP:  52%|█████▏    | 13/25 [17:44<18:08, 90.74s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:31, 91.83s/it]               
computing SHAP:  56%|█████▌    | 14/25 [19:16<16:41, 91.07s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:18, 78.71s/it]               
computing SHAP:  60%|██████    | 15/25 [20:35<14:33, 87.35s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:38, 98.99s/it]               
computing SHAP:  64%|██████▍   | 16/25 [22:14<13:37, 90.85s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:28, 88.65s/it]               
computing SHAP:  68%|██████▊   | 17/25 [23:42<12:01, 90.19s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:47, 107.86s/it]              
computing SHAP:  72%|███████▏  | 18/25 [25:30<11:08, 95.51s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:02, 62.83s/it]               
computing SHAP:  76%|███████▌  | 19/25 [26:33<08:34, 85.69s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:29, 89.83s/it]               
computing SHAP:  80%|████████  | 20/25 [28:03<07:14, 86.94s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:40, 100.14s/it]              
computing SHAP:  84%|████████▍ | 21/25 [29:43<06:03, 90.91s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:03, 63.22s/it]               
computing SHAP:  88%|████████▊ | 22/25 [30:46<04:07, 82.60s/it]

  0%|          | 0/72 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [00:55, 55.63s/it]               
computing SHAP:  92%|█████████▏| 23/25 [31:42<02:29, 74.51s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:23, 83.27s/it]               
computing SHAP:  96%|█████████▌| 24/25 [33:05<01:17, 77.14s/it]

  0%|          | 0/98 [00:00<?, ?it/s]


PartitionExplainer explainer: 2it [01:09, 69.12s/it]               
computing SHAP: 100%|██████████| 25/25 [34:14<00:00, 82.19s/it]


  explainability: 0.6048  |  mean gini: 0.6048  |  avg top tokens: 2.0

done. evaluated 5 models.


In [6]:
with open("explainability_scores.json", "w") as f:
    json.dump({"explainability": results}, f, indent=2)
print("saved to explainability_scores.json")

saved to explainability_scores.json


**Conclusion**
- **Qwen2-0.5B is the most explainable** (0.64 Gini), concentrating attribution on the fewest tokens (1.28 avg), meaning its predictions are driven by the sharpest, most focused signals.

- **TinyLlama is the least explainable by far** (0.33 Gini), nearly half the score of the others, spreading attribution almost uniformly across tokens rather than focusing on meaningful ones.

- **Phi-1.5, SmolLM, and StableLM cluster tightly** (0.60 to 0.64), showing broadly similar explainability despite different architectures and sizes, suggesting model size alone does not determine attribution concentration.

- **StableLM relies on the most tokens per prediction** (2.0 avg top tokens), meaning even with a competitive Gini score it distributes importance across more tokens than the others, making individual predictions harder to interpret.

**Next: `06_privacy_score.ipynb`**

Measuring privacy risk via MIA canary insertion test and PII generation risk.

Does the model memorise and reproduce sensitive training data?